javaslang/src/main/java/javaslang/collection/PriorityQueue.java

```diff
@@ -0,0 +1,780 @@
+         * *  else                                                   Node (x0,r1+1,[t1, t2])
+         **/
+        private static <T> Node<Ranked<T>> skewLink(SerializableComparator<? super T> comparator, Node<Ranked<T>> tree0, Node<Ranked<T>> tree1, Node<Ranked<T>> tree2) {
+            final T root0 = root(tree0), root1 = root(tree1), root2 = root(tree2);
+
+            final Ranked<T> ranked;
+            final List<Node<Ranked<T>>> children;
+            if (comparator.isLessOrEqual(root1, root0) && comparator.isLessOrEqual(root1, root2)) {
+                ranked = Ranked.of(root1, rank(tree1) + 1);
+                children = concat(tree0, concat(tree2, tree1.getChildren()));
+            } else if (comparator.isLessOrEqual(root2, root0) && comparator.isLessOrEqual(root2, root1)) {
+                ranked = Ranked.of(root2, rank(tree2) + 1);
+                children = concat(tree0, concat(tree1, tree2.getChildren()));
+            } else {
+                ranked = Ranked.of(root0, rank(tree1) + 1);
+                children = List.of(tree1, tree2);
+            }
+            return Tree.of(ranked, children);
+        }
+
+        /**
+         * fun findMin [] = raise EMPTY
+         * * | findMin [t] = root t
+         * * | findMin (t :: ts) =
+         * *     let val x = findMin ts
+         * *     in if Elem.leq (root t, x) then root t else x end
+         */
+        protected static <T> T findMin(SerializableComparator<? super T> comparator, List<Node<Ranked<T>>> forest, T currentMin) {
+            while (forest.size() > 1) {
+                forest = forest.tail();
+                currentMin = min(comparator, currentMin, root(forest.head()));
+            }
+            return currentMin;
+        }
+
+        private static <T> T min(SerializableComparator<? super T> comparator, T value1, T value2) {
+            return comparator.isLessOrEqual(value1, value2) ? value1 : value2;
+        }
+
+        protected static <T> T root(Node<Ranked<T>> tree) {
+            return tree.getValue().value;
+        }
+
+        private static <T> int rank(Node<Ranked<T>> tree) {
+            return tree.getValue().rank;
+        }
+
+        private static <T> List<T> concat(T head, List<T> rest) {
+            return rest.prepend(head);
+        }
+
+        protected static class Ranked<T> implements Serializable {
+            private static final long serialVersionUID = 1L;
+
+            final T value;
+            final int rank;
+
+            Ranked(T value, int rank) {
+                this.value = value;
+                this.rank = rank;
+            }
+
+            static <T> Ranked<T> of(T value, int rank) {
+                return new Ranked<>(value, rank);
+            }
+
+            @Override
+            public String toString() {
+                return "Ranked(" + value + ", " + rank + ')';
+            }
+        }
+    }
+
+    /* TODO decide whether we should switch to this impl if we have few elements */
```

Comment:
I think regarding PriorityQueues, the mutable (unsynchronized) form is unbeaten. If there is a scenario where the Queue is the bottleneck I would suggest to use the mutable one.

I love your purely functional implementation and want to keep it. We should remove the delegate.